### ISSUE TITLE
Overeager "reportUnnecessaryComparison" when comparing set and frozenset

### ISSUE THREAD

**Issue by Hoolean** [microsoft/pyright#4423]
**Describe the bug**

pyright expects comparisons between `set()` and `frozenset()` to always evaluate to False, because their types have no overlap. In some cases, however, Python will consider them equal, as their equality is based on whether their contained values are the same, irrespective of whether they are frozen.

**Reproducing code**

```python
# pyright expects the following condition to always evaluate to False, when in
# this case it evaluates to True.
print({1, 2, 3} == frozenset({1, 2, 3}))  # True
```

**Expected behavior**

Either:
1. pyright does not flag "reportUnnecessaryComparison", as the comparison's result is dependent on the sets' values; or
2. pyright is able to identify that the sets' values are static and equal, and that the expression will always evaluate to True.

**Actual behavior**

pyright flags "reportUnnecessaryComparison", believing that the comparison will always evaluate to False.

**VS Code extension or command-line**

| | Version |
|-|-|
| VS Code | `1.73.1` |
| Pylance | `v2023.1.10` |

**Comment by erictraut** [comment]
I was reluctant to implement the `reportUnnecessaryComparison` check because it had the potential for false positives like this. There's no good way for a static type checker to know, for example, that the `frozenset` implementation allows comparison with a `set`. It's not feasible to hard-code all of these internal implementation peculiarities.

Ultimately, we concluded that this is a strict-mode only test, so it's OK if it has some rare false positives. You can choose to disable it globally or on a case-by-case basis if you'd like.

**Comment by Hoolean** [comment]
Understood, thanks for taking a look Eric :)
